src/Equations/EffectiveCriticalField.cpp

```diff
@@ -0,0 +1,396 @@
+    : Eceff_mode(par->Eceff_mode), collSettingsForEc(par->collSettingsForEc), collQtySettings(par->collQtySettings), 
+    rGrid(par->rGrid), nuS(par->nuS), nuD(par->nuD), ions(par->ions), lnLambda(par->lnLambda), 
+    fsolve(par->fsolve)
+    {
+        gsl_parameters.rGrid = par->rGrid;
+        gsl_parameters.nuS = par->nuS;
+        gsl_parameters.nuD = par->nuD;
+        gsl_parameters.fgType = par->fgType;
+        gsl_parameters.gsl_ad_w = par->gsl_ad_w;
+        gsl_parameters.fmin = par->fmin;
+        gsl_parameters.collSettingsForEc = par->collSettingsForEc;
+        gsl_parameters.QAG_KEY = GSL_INTEG_GAUSS31;
+        gsl_parameters.analyticDist = new AnalyticDistribution(rGrid, nuD, Eceff_mode);
+    }
+
+EffectiveCriticalField::~EffectiveCriticalField(){
+    delete gsl_parameters.analyticDist;
+}
+
+
+/**
+ * Calculates and stores the effective critical field for runaway generation.
+ * The calculation is based on Eq (21) in Hesslow et al, PPCF 60, 074010 (2018)
+ * but has been generalized to account for inhomogeneous magnetic fields. The
+ * method implemented here is outlined in DREAM/doc/notes/theory.
+ * Essentially, the critical effective electric field is defined as the E
+ * for which the maximum (with respect to p) of U(p) equals 0. Here, U
+ * is the net momentum advection term averaged over an analytic pitch
+ * angle distribution.
+ */
+void EffectiveCriticalField::CalculateEffectiveCriticalField(const real_t *Ec_tot, const real_t *Ec_free, real_t *effectiveCriticalField){
+    len_t nr = rGrid->GetNr();
+    switch (Eceff_mode)
+    {
+        case OptionConstants::COLLQTY_ECEFF_MODE_EC_TOT : { // or COLLQTY_ECEFF_MODE_NOSCREENING to be consistent with 
+        // for example GetConnorHastieField_NOSCREENING?
+            if(collQtySettings->collfreq_type==OptionConstants::COLLQTY_COLLISION_FREQUENCY_TYPE_COMPLETELY_SCREENED){
+                for(len_t ir=0; ir<nr; ir++)
+                    effectiveCriticalField[ir] = Ec_free[ir];
+            }else{
+                for(len_t ir=0; ir<nr; ir++)
+                    effectiveCriticalField[ir] = Ec_tot[ir]; 
+            }
+        } 
+        break;
+        case OptionConstants::COLLQTY_ECEFF_MODE_CYLINDRICAL : {
+            // or should we use CalculateEceffPPCFPaper regardless, to account for radiation reaction? 
+            // I think it is more accurate than Ec_free at least. 
+            if(collQtySettings->collfreq_type==OptionConstants::COLLQTY_COLLISION_FREQUENCY_TYPE_COMPLETELY_SCREENED){
+                for(len_t ir=0; ir<nr; ir++)
+                    effectiveCriticalField[ir] = Ec_free[ir];  
+            }else{
+                // calculate Eceff according to paper
+                for(len_t ir=0; ir<nr; ir++){
+                    effectiveCriticalField[ir] = CalculateEceffPPCFPaper(ir);
+                }
+            }
+        }
+        break;
+        case OptionConstants::COLLQTY_ECEFF_MODE_SIMPLE : 
+        {
```

Comment:
actually, how does this work? @hesslow @hoppe93 
does this cause SIMPLE to fallthrough to FULL (which is the desired behaviour)?

Reply:
No, I would expect this to be invalid code. The case will fallthrough as long as there is no ``break`` between them. In this mode, it looks as if the ``FULL`` case is within the ``SIMPLE`` case, and thus not accessible to the switch statement? (Probably it is though, because otherwise I think the compiler should've complained)

Reply:
I changed this to something that is hopefully clearer now

Reply:
Hm, seems like I missed the C++17 updates for switch statements (if I understood correctly)? Thanks for fixing!